core/src/main/scala/org/apache/spark/status/api/v1/EventLogDownloadResource.scala

```diff
@@ -0,0 +1,110 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.status.api.v1
+
+import java.io.{File, FileOutputStream, OutputStream}
+import javax.ws.rs.{GET, Produces}
+import javax.ws.rs.core.{MediaType, Response, StreamingOutput}
+
+import scala.collection.mutable.ArrayBuffer
+
+import org.apache.hadoop.fs.Path
+
+import org.apache.spark.{Logging, SparkConf}
+import org.apache.spark.deploy.SparkHadoopUtil
+import org.apache.spark.deploy.history.HistoryServer
+import org.apache.spark.util.Utils
+
+@Produces(Array(MediaType.APPLICATION_OCTET_STREAM))
+private[v1] class EventLogDownloadResource(
+    val uIRoot: UIRoot,
+    val appId: String,
+    val attemptId: Option[String]) extends Logging {
+  val conf = SparkHadoopUtil.get.newConfiguration(new SparkConf)
+
+  @GET
+  def getEventLogs(): Response = {
+    uIRoot match {
+      case hs: HistoryServer =>
+        var logsNotFound = false
+        val fileName = {
+          attemptId match {
+            case Some(id) => s"eventLogs-$appId-$id.zip"
+            case None => s"eventLogs-$appId.zip"
+          }
+        }
+        val stream = new StreamingOutput {
+          override def write(output: OutputStream): Unit = {
+            val eventLogs = hs.getEventLogPaths(appId, attemptId)
+            if (eventLogs.isEmpty) logsNotFound = true
+            else zipLogFiles(eventLogs, output)
+          }
+        }
+        if (logsNotFound) {
+          Response.serverError()
+            .entity(s"Event logs are not available for app: $appId.")
+            .status(Response.Status.SERVICE_UNAVAILABLE)
+            .build()
+        } else {
+          Response.ok(stream)
+            .header("Content-Disposition", s"attachment; filename=$fileName")
+            .header("Content-Type", MediaType.APPLICATION_OCTET_STREAM)
+            .build()
+        }
+      case _ =>
+        Response.serverError()
+          .entity("Event logs are only available through the history server.")
+          .status(Response.Status.SERVICE_UNAVAILABLE)
+          .build()
+    }
+  }
+
+  private def zipLogFiles(eventLogs: Seq[Path], output: OutputStream): Unit = {
+    val areLegacyLogs = eventLogs.headOption.exists { path =>
+      path.getFileSystem(conf).isDirectory(path)
+    }
+    val pathsToZip = if (areLegacyLogs) {
+      new ArrayBuffer[Path]()
+    } else {
+      eventLogs
+    }
+    var tempDir: File = null
+    try {
+      if (areLegacyLogs) {
+        tempDir = Utils.createTempDir()
+        Utils.chmod700(tempDir)
+        eventLogs.foreach { logPath =>
+          // If the event logs are directories (legacy), then create a zip file for each
```

Comment:
The way Java handles directory writes to zip archive is weird at best - there is no direct way to handle directory writes, but by adding entry names as dirName/pathName. This would mean that for each entry we'd have to parse the directory name. And if there are several levels our code will need to handle that (considering the zipping code is in Utils - so it will need to be generic enough to handle any number of levels. I simply chose to not support that). Plus since this is for legacy directories, this code path is going to be exercised pretty rarely. I'd just leave this as is to keep the code simple.